### PR TITLE
Release 24.1.0

### PR DESCRIPTION
**Pull request by rumchoto**

### CODE REVIEW

**Review by marcosjgutierrez — COMMENTED** [review]
Non-blocking comments. Mainly related to documentation, error handling and hardcoded values. 

Comments follow https://conventionalcomments.org/

**Review by marcosjgutierrez — APPROVED** [review]
Great job on this PR! I've left some non-blocking comments about error handling, documentation, and hardcoded values. While I'm approving this PR, please consider addressing these points in a future commit for code quality improvements.

File: cartridges/int_klarna_payments/cartridge/static/default/js/klarna-payments.js [hunk 1]
```diff
@@ -35,6 +35,7 @@
         } );
 
         var $selectedPaymentMethod = $selectPaymentMethod.querySelectorAll( ':checked' )[0];
+		console.log('klarnaAsyncCallback');
```
[the remark this answers is not in the document]
Comment (rumchoto) [reply cartridges/int_klarna_payments/cartridge/static/default/js/klarna-payments.js:38]: no, it is not required. Removed.

File: cartridges/int_klarna_payments_controllers/cartridge/controllers/KlarnaPayments.js [hunk 2]
```diff
@@ -875,42 +896,358 @@ function cancelSubscription() {
     return;
 }
 
+function handleAuthorizationResult() {
```
[the remark this answers is not in the document]
Comment (rumchoto) [reply cartridges/int_klarna_payments_controllers/cartridge/controllers/KlarnaPayments.js:899]: Comments are added.

File: cartridges/int_klarna_payments_sfra/cartridge/client/default/js/klarnaExpressCheckout.js [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+    if ($('.ec-payment-failure-alert').length === 0) {
+        $('.ec-payment-failure').append(
+            '<div class="alert ' + 'alert-danger' + ' ec-payment-failure-alert add-to-basket-alert text-center" role="alert">'
+            + msg + '</div>'
+        );
+    }
+
+    setTimeout(function () {
+        $('.ec-payment-failure-alert').remove();
+    }, 2000);
```
[the remark this answers is not in the document]
Comment (rumchoto) [reply cartridges/int_klarna_payments_sfra/cartridge/client/default/js/klarnaExpressCheckout.js:26]: Values are moved to constants.

File: cartridges/int_klarna_payments_sfra/cartridge/client/default/js/klarnaExpressCheckout.js [hunk 4]
```diff
@@ -0,0 +1,171 @@
...
+                                            success: function () {
+                                                setTimeout(function () {
+                                                    generateErrorAuthMessage();
+                                                }, 100);
+                                            }
+                                        });
+                                    } else {
+                                        setTimeout(function () {
+                                            generateErrorAuthMessage();
+                                        }, 100);
```
[the remark this answers is not in the document]
Comment (rumchoto) [reply cartridges/int_klarna_payments_sfra/cartridge/client/default/js/klarnaExpressCheckout.js:130]: Values are moved to constants.

File: cartridges/int_klarna_payments_sfra/cartridge/client/default/js/klarnaExpressCheckout.js [hunk 5]
```diff
@@ -0,0 +1,171 @@
...
+                                } else {
+                                    //revert original basket
+                                    if (isPDP) {
+                                        $.ajax({
+                                            url: klarnaPaymentsUrls.handleAuthFailurePDP,
+                                            type: 'get',
+                                            dataType: 'json',
+                                            contentType: 'application/json',
+                                            success: function () {
+                                                setTimeout(function () {
```
[the remark this answers is not in the document]
Comment (rumchoto) [reply cartridges/int_klarna_payments_sfra/cartridge/client/default/js/klarnaExpressCheckout.js:122]: done.

File: cartridges/int_klarna_payments_sfra/cartridge/client/default/js/klarnaExpressCheckoutPDP.js [hunk 6]
```diff
@@ -0,0 +1,21 @@
+
+'use strict';
+
+$(function () {
```
[the remark this answers is not in the document]
Comment (rumchoto) [reply cartridges/int_klarna_payments_sfra/cartridge/client/default/js/klarnaExpressCheckoutPDP.js:4]: documented

File: cartridges/int_klarna_payments_sfra/cartridge/client/default/js/klarnaPayments.js [hunk 7]
```diff
@@ -1141,3 +1164,40 @@ window.klarnaAsyncCallback = function () {
...
     });
 };
+
+/**
+ * In case of error fields in the inoput form,
+ * the form is displayed for correction
+ * @param {object} $form checkout form to check for errors
+ */
+function checkFormErrors($form) {
+    var numOfTries = 3;
```
[the remark this answers is not in the document]
Comment (rumchoto) [reply cartridges/int_klarna_payments_sfra/cartridge/client/default/js/klarnaPayments.js:1174]: Values are moved to constants.

File: cartridges/int_klarna_payments_sfra/cartridge/client/default/js/klarnaPayments.js [hunk 8]
```diff
@@ -1141,3 +1164,40 @@ window.klarnaAsyncCallback = function () {
...
+            clearInterval(interval);
+            return;
+        }
+        numOfTries--;
+        var errors = $form.find('.form-control.is-invalid');
+        if (errors.length) {
+            $('.btn-show-details').trigger('click');
+            clearInterval(interval);
+        }
+    }, 100);
```
[the remark this answers is not in the document]
Comment (rumchoto) [reply cartridges/int_klarna_payments_sfra/cartridge/client/default/js/klarnaPayments.js:1186]: Values are moved to constants.

File: cartridges/int_klarna_payments_sfra/cartridge/controllers/KlarnaPayments.js [hunk 9]
```diff
@@ -455,4 +455,316 @@ server.get('CancelSubscription', userLoggedIn.validateLoggedInAjax, function (re
     return next();
 });
 
+server.post('ECAuthorizationCallback', function (req, res, next) {
+    var Transaction = require('dw/system/Transaction');
+    // var klarnaResponse = JSON.parse(req.body);
+    dw.system.Logger.error(req.body);
+    Transaction.wrap(function () {
+        //TODO
```
[the remark this answers is not in the document]
Comment (rumchoto) [reply cartridges/int_klarna_payments_sfra/cartridge/controllers/KlarnaPayments.js:463]: Comments added.

File: cartridges/int_klarna_payments_sfra/cartridge/controllers/KlarnaPayments.js [hunk 10]
```diff
@@ -455,4 +455,316 @@ server.get('CancelSubscription', userLoggedIn.validateLoggedInAjax, function (re
     return next();
 });
 
+server.post('ECAuthorizationCallback', function (req, res, next) {
+    var Transaction = require('dw/system/Transaction');
+    // var klarnaResponse = JSON.parse(req.body);
+    dw.system.Logger.error(req.body);
```
[the remark this answers is not in the document]
Comment (rumchoto) [reply cartridges/int_klarna_payments_sfra/cartridge/controllers/KlarnaPayments.js:461]: refactored

File: cartridges/int_klarna_payments_sfra/cartridge/models/product/decorators/base.js [hunk 11]
```diff
@@ -0,0 +1,39 @@
+'use strict';
+
+function isSubscriptionOnly(apiProduct) {
```
[the remark this answers is not in the document]
Comment (rumchoto) [reply cartridges/int_klarna_payments_sfra/cartridge/models/product/decorators/base.js:3]: commented

File: cartridges/int_klarna_payments_sfra/cartridge/scripts/util/klarnaHelper.js [hunk 12]
```diff
@@ -145,6 +157,45 @@ function filterApplicableShippingMethods(shipment, address) {
...
+superMdl.revertCurrentBasketProductData = function (currentBasket) {
+    var Transaction = require('dw/system/Transaction');
+    var cartHelper = require('*/cartridge/scripts/cart/cartHelpers');
+    var basketCalculationHelpers = require('*/cartridge/scripts/helpers/basketCalculationHelpers');
+
+    if (!empty(currentBasket)) {
+        var customerBasketData = session.privacy.kpCustomerProductData ? JSON.parse(session.privacy.kpCustomerProductData) : null;
+        var productLineItems = currentBasket.productLineItems.toArray();
+
+        Transaction.wrap(function () {
```
[the remark this answers is not in the document]
Comment (rumchoto) [reply cartridges/int_klarna_payments_sfra/cartridge/scripts/util/klarnaHelper.js:170]: done

File: cartridges/int_klarna_payments_sfra/cartridge/templates/default/components/footer/pageFooter.isml [hunk 13]
```diff
@@ -42,3 +43,17 @@
...
+        <isinclude template="components/modules" />
+        <isinclude template="klarnapayments/modules.isml" />
+        <iskecminicart />
+    </div>
+    <script>
+            <isinclude template="/resources/klarnapaymentsresources.isml"/>
+        </script>
+    <script type="text/javascript" src="${URLUtils.staticURL('/js/klarnaExpressCheckout.js')}"></script>
+    <link rel="stylesheet" href="${URLUtils.staticURL('/css/klarnaExpress.css')}" />
+    <script async src="https://x.klarnacdn.net/kp/lib/v1/api.js"></script>
```
[the remark this answers is not in the document]
Comment (rumchoto) [reply cartridges/int_klarna_payments_sfra/cartridge/templates/default/components/footer/pageFooter.isml:58]: Script URL is moved to site preference.

File: cartridges/int_klarna_payments/cartridge/templates/default/klarnapayments/scripts.isml [hunk 14]
```diff
@@ -17,4 +17,16 @@
...
+
+<isif condition="${KlarnaOSM.isKlarnExpressCheckoutEnabled() && KlarnaOSM.getKlarnExpressCheckoutClientKey()}">
+
+    <script>
+		<isinclude template="/resources/klarnapaymentsresources.isml"/>
+	</script>
+    <script type="text/javascript" src="${URLUtils.staticURL('/js/klarna-express-checkout.js')}"></script>
+    <link rel="stylesheet" href="${URLUtils.staticURL('/css/klarna-express.css')}" />
+    <isif condition="${pageContext.ns !== 'checkout'}">
+        <script async src="https://x.klarnacdn.net/kp/lib/v1/api.js"></script>
```
[the remark this answers is not in the document]
Comment (rumchoto) [reply cartridges/int_klarna_payments/cartridge/templates/default/klarnapayments/scripts.isml:30]: Script URL is moved to site preference.

**Pull request merged**